### PR TITLE
Add /_matrix/versions to report supported versions

### PR DESCRIPTION
**Pull request by illicitonion**

### CODE REVIEW

File: synapse/rest/client/versions.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+logger = logging.getLogger(__name__)
+
+
+class VersionsRestServlet(RestServlet):
+    PATTERNS = [re.compile("^/_matrix/versions$")]
+
+    def on_GET(self, request):
+        return (200, {
+            "versions": [
+                "r0.0.1",
```
Comment (richvdh) [inline synapse/rest/client/versions.py]: so, what is r0.0.1? Hard to claim we support it before we've defined it.

Comment (illicitonion) [inline synapse/rest/client/versions.py]: The thing I'm about to cut as soon as https://github.com/matrix-org/matrix-doc/pull/250 lands :)

File: synapse/rest/client/versions.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+from synapse.http.servlet import RestServlet
+
+import logging
+import re
+
+logger = logging.getLogger(__name__)
+
+
+class VersionsRestServlet(RestServlet):
+    PATTERNS = [re.compile("^/_matrix/versions$")]
```
Comment (richvdh) [inline synapse/rest/client/versions.py]: needs a change now

File: synapse/app/homeserver.py [hunk 3]
```diff
@@ -185,6 +185,7 @@ def _listener_http(self, config, listener_config):
                         "/_matrix/client/r0": client_resource,
                         "/_matrix/client/unstable": client_resource,
                         "/_matrix/client/v2_alpha": client_resource,
+                        "/_matrix/versions": client_resource,
```
Comment (richvdh) [inline synapse/app/homeserver.py]: this needs a change.

Having this list is *horrible*, though I realise fixing it is probably out of scope right now.

File: synapse/rest/client/versions.py (hunk 2 above)
Comment (illicitonion) [inline synapse/rest/client/versions.py]: Done

File: synapse/app/homeserver.py (hunk 3 above)
Comment (illicitonion) [inline synapse/app/homeserver.py]: Done

I was also unhappy to find this. In particular, the digging through when all I knew was that the thing 404'd, was not fun.

**Comment by richvdh** [comment]
LGTM.

**Pull request merged**